sdks/python/apache_beam/runners/worker/worker_status.py

```
@@ -250,6 +263,13 @@ def _log_lull_in_bundle_processor(self, bundle_process_cache):
           if processor:
             info = processor.state_sampler.get_info()
             self._log_lull_sampler_info(info, instruction)
+            if self._restart_lull_timeout_ns and self._restart_lull(info):
+              sys.exit(1)
```

On the design... we want to be very careful about adding `exit` calls (here and in Java). Because eventually code ends up getting used in unexpected contexts. It should be written without assumptions about the context it is called in, to the extent this is possible. So this could cause some other way of executing to exit unexpectedly. Can we raise an exception that bubbles up and causes the top level `main` to exit? That is really the only level where we should make any assumptions about being in charge of the process lifetime.

---

+1, the exceptions route aligns well with how the timeout was implemented in the Java SDK.

---

Modified